### PR TITLE
adapt updateContainer interface change

### PR DESCRIPTION
**Pull request by SteffenSeckler**
# Description

Adapts change of updateContainer function, i.e., they now return a vector of invalid particles.

## Related Pull Requests

- part of #215

### CODE REVIEW

File: src/autopas/utils/AutoPasMacros.h [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+ * @date 14.05.19
+ */
+
+#pragma once
+
+#ifdef __GNUC__
+/**
+ * @todo c++17: __attribute__((warn_unused_result)) should be replaced with [[nodiscard]]
+ */
+#define AUTOPAS_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
```
Comment (FG-TUM) [inline src/autopas/utils/AutoPasMacros.h]: clang version?

File: src/autopas/containers/linkedCells/LinkedCells.h [hunk 2]
```diff
@@ -116,15 +116,11 @@ class LinkedCells : public ParticleContainer<Particle, ParticleCell, SoAArraysTy
...
-      utils::ExceptionHandler::exception(
-          "Linked Cells: Halo particles still present when updateContainer was called. First particle found:\n" +
-          haloIter->toString());
-    }
-
+  std::vector<Particle> AUTOPAS_WARN_UNUSED_RESULT updateContainer() override {
+    this->deleteHaloParticles();
+    std::vector<Particle> invalidParticles;
 #ifdef AUTOPAS_OPENMP
+#pragma omp declare reduction(vecMerge : std::vector<Particle> : omp_out.insert(omp_out.end(), omp_in.begin(), omp_in.end()))
```
Comment (FG-TUM) [inline src/autopas/containers/linkedCells/LinkedCells.h]: Move to WrapOpenMP.h

File: src/autopas/containers/verletClusterLists/VerletClusterLists.h [hunk 3]
```diff
@@ -119,15 +122,33 @@ class VerletClusterLists : public ParticleContainer<Particle, FullParticleCell<P
    * @copydoc VerletLists::deleteHaloParticles
    */
   void deleteHaloParticles() override {
-    autopas::utils::ExceptionHandler::exception("VerletClusterLists.deleteHaloParticles not yet implemented.");
+    // quick and dirty: iterate over all particles and delete halo particles
```
Comment (FG-TUM) [inline src/autopas/containers/verletClusterLists/VerletClusterLists.h]: add a todo and maybe an issue

File: src/autopas/containers/verletClusterLists/VerletClusterLists.h [hunk 4]
```diff
@@ -119,15 +122,33 @@ class VerletClusterLists : public ParticleContainer<Particle, FullParticleCell<P
...
    */
-  void updateContainer() override {
+  std::vector<Particle> AUTOPAS_WARN_UNUSED_RESULT updateContainer() override {
     AutoPasLog(debug, "updating container");
+    // first delete all particles
+    this->deleteHaloParticles();
+
+    // next find invalid particles
+    std::vector<Particle> invalidParticles;
+    for (auto iter = this->begin(IteratorBehavior::ownedOnly); iter.isValid(); ++iter) {
```
Comment (FG-TUM) [inline src/autopas/containers/verletClusterLists/VerletClusterLists.h]: is this parallelizable?

File: src/autopas/particles/Particle.h [hunk 5]
```diff
@@ -196,6 +198,11 @@ class ParticleBase {
    * Particle id.
    */
   unsigned long _id;
+
+  /**
+   * Defines whether the particle is owned by the current process
```
Comment (FG-TUM) [inline src/autopas/particles/Particle.h]:
```suggestion
   * Defines whether the particle is owned by the current (MPI-)process
```

File: src/autopas/containers/ParticleContainerInterface.h [hunk 6]
```diff
@@ -145,7 +146,7 @@ class ParticleContainerInterface {
    * This resorts particles into appropriate cells and moves them to the halo, if necessary.
    * @todo: update doc
```
Comment (FG-TUM) [inline src/autopas/containers/ParticleContainerInterface.h]: !!!

Comment (SteffenSeckler) [reply src/autopas/containers/ParticleContainerInterface.h]: later

File: src/autopas/containers/verletClusterLists/VerletClusterLists.h (hunk 4 above)
Comment (SteffenSeckler) [reply src/autopas/containers/verletClusterLists/VerletClusterLists.h]: added todo in code + #215

File: src/autopas/containers/linkedCells/LinkedCells.h (hunk 2 above)
Comment (SteffenSeckler) [reply src/autopas/containers/linkedCells/LinkedCells.h]: not possible, as it uses a template

Comment (SteffenSeckler) [reply src/autopas/containers/linkedCells/LinkedCells.h]: actually, this isn't possible at all. now using critical sections for this...

File: src/autopas/utils/AutoPasMacros.h (hunk 1 above)
Comment (SteffenSeckler) [reply src/autopas/utils/AutoPasMacros.h]: clang apparently also defines __GNUC__, but added __clang__ anyway. 
also tested it. __attribute__((warn_unused_result)) works with clang as well.

**Pull request merged**